src/ring_theory/graded_algebra/homogeneous_ideal.lean

```diff
@@ -0,0 +1,493 @@
+/-
+Copyright (c) 2021 Jujian Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jujian Zhang, Eric Wieser
+-/
+
+import ring_theory.ideal.basic
+import ring_theory.ideal.operations
+import linear_algebra.finsupp
+import ring_theory.graded_algebra.basic
+
+/-!
+
+# Homogeneous ideal of a graded algebra
+
+This file defines homogeneous ideals of `graded_algebra A` where `A : ι → ideal R`and operations on
+them:
+* `mul`, `inf`, `Inf` of homogeneous ideals are homogeneous;
+* `⊤`, `⊥`, i.e. the trivial ring and `R` are homogeneous;
+* `radical` of a homogeneous ideal is homogeneous.
+-/
+
+open set_like direct_sum set
+open_locale big_operators pointwise direct_sum
+
+section homogeneous_core
+
+variables {ι R A : Type*} [comm_ring R] [ring A] [algebra R A]
+variables (𝒜 : ι → submodule R A)
+variable (I : ideal A)
+
+/-- For any `I : ideal R`, not necessarily homogeneous, there is a homogeneous ideal associated with
+`I` spanned by all homogeneous elements in `I`. This construction is used when proving that the
+radical of a homogeneous ideal is homogeneous. -/
+def ideal.homogeneous_core : ideal A :=
+ideal.span (coe '' ((coe : subtype (is_homogeneous 𝒜) → A) ⁻¹' I))
+
+lemma ideal.homogeneous_core_is_mono : monotone (ideal.homogeneous_core 𝒜) :=
+λ I J I_le_J, ideal.span_mono $ set.image_subset _ $ λ x, @I_le_J _
+
+lemma ideal.homogeneous_core_le_ideal : ideal.homogeneous_core 𝒜 I ≤ I :=
+ideal.span_le.2 $ image_preimage_subset _ _
+
+end homogeneous_core
+
+section is_homogeneous_ideal_defs
+
+variables {ι R A : Type*} [comm_ring R] [ring A] [algebra R A]
+variables (𝒜 : ι → submodule R A)
+variable (I : ideal A)
+
+lemma ideal.is_homogeneous.exists_subset_iff_eq_span :
+  (∃ s : set A, s ⊆ set_of (is_homogeneous 𝒜) ∧ I = ideal.span s) ↔
+    I = ideal.span (I ∩ set_of (is_homogeneous 𝒜)) :=
+begin
+  split,
+  { rintros ⟨s, hs, rfl⟩,
+    apply le_antisymm,
+    { exact ideal.span_mono (subset_inter ideal.subset_span hs) },
+    { exact ideal.span_le.2 (inter_subset_left _ _) } },
+  { intros hI,
+    exact ⟨_, inter_subset_right _ _, hI⟩, }
+end
+
+variables [decidable_eq ι] [add_comm_monoid ι]  [graded_algebra 𝒜]
+
+/--An `I : ideal R` is homogeneous if for every `r ∈ I`, all homogeneous components
+  of `r` are in `I`.-/
+def ideal.is_homogeneous : Prop :=
+∀ (i : ι) ⦃r : A⦄, r ∈ I → (graded_algebra.decompose 𝒜 r i : A) ∈ I
+
+lemma ideal.is_homogeneous_iff_forall_subset :
+  ideal.is_homogeneous 𝒜 I ↔ ∀ i, (I : set A) ⊆ graded_algebra.proj 𝒜 i ⁻¹' I :=
+iff.rfl
+
+lemma ideal.is_homogeneous_iff_subset_Inter :
+  ideal.is_homogeneous 𝒜 I ↔ (I : set A) ⊆ ⋂ i, graded_algebra.proj 𝒜 i ⁻¹' ↑I :=
+subset_Inter_iff.symm
+
+lemma ideal.is_homogeneous.exists_iff_eq_span :
+  (∃ (S : set (homogeneous_submonoid 𝒜)), I = ideal.span (coe '' S)) ↔
+  I = ideal.span {x | x ∈ I ∧ is_homogeneous 𝒜 x} :=
+(subtype.exists_set_subtype _).trans (ideal.is_homogeneous.exists_subset_iff_eq_span _ _)
```

Comment:
I think this lemma should be changed to:
```suggestion
lemma ideal.is_homogeneous.exists_iff_eq_span :
  (∃ (S : set (homogeneous_submonoid 𝒜)), I = ideal.span (coe '' S)) ↔
    I = ideal.homogeneous_core 𝒜 I :=
(set.image_preimage.compose (submodule.gi _ _).gc).exists_eq_l _
```
and actually, we should just remove `ideal.is_homogeneous.exists_subset_iff_eq_span` entirely (sorry for suggesting that we add it!)